lib/spack/spack/environment/environment.py

```diff
@@ -1618,11 +1620,10 @@ def matching_spec(self, spec):
         for user_spec, concretized_user_spec in self.concretized_specs():
             # Deal with concrete specs differently
             if spec.concrete:
-                # Matching a concrete spec is more restrictive
-                # than just matching the dag hash
+                # TODO: do we still need the extra check comparing dag hashes?
```

Comment:
If we are using the `full_hash` for `Spec.__contains__` (which we probably should?) then the comparison seems redundant

Reply:
The `full_hash` is gone now, or at least only used when reading old spec dicts, so I'm taking this comment to mean `dag_hash`.  Anyway, based on a quick glance at `Spec.__contains__`, checking a hash there is too specific/constraining.  To me `__contains__` *appears* to be about whether its object simply satisfies its argument.  Looking at this again today, I see we only did the hash comparison if the spec we were given (the one for which we wish to find matches) was concrete.  Since all the match candidates must be concrete, I think hash comparison is fine here, because how could one concrete spec "satisfy" another if they aren't the same?  I think I'll just remove the `TODO` comment and keep the hash comparison, but let me know if you disagree?

Reply:
I'd resolve this now, but want to allow a chance for discussion.

Reply:
Yep this should no longer be needed.  We check if it's concrete first;  the only way it'll be in the `concretized_user_spec` is if its `dag_hash` matches something there.